### PR TITLE
Ref must come before 'partial' in a 'ref partial struct'

### LINKED ISSUE #22379: Parse "ref" before "partial" in "ref partial struct"
Currently, the parser expects "partial ref struct". But from LDM discussion, we'd like to treat `ref` more like a modifier in this context, and possibly allow it in any order in the bag of modifiers.
In the meantime, we can open the door for this later change by parsing "ref partial struct".

### PR DESCRIPTION
**Pull request by agocke**
**Customer scenario**

The previous LDM design for `ref` and `partial` on structs is that `partial` comes before `ref`. This implements the change to reverse that.

**Bugs this fixes:**

Fixes #22379 

**Risk**

This change is isolated to parsing `ref` as a modifier on structs.

**Performance impact**

None. No major new code, same parsing complexity.

**Root cause analysis:**

LDM change.

### CODE REVIEW

**Comment by agocke** [comment]
This PR has been updated to also fix a bug that allowed certain modifiers to appear after `partial` but before a type declaration keyword in a type declaration.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 1]
```diff
@@ -336,9 +337,24 @@ private DeclarationModifiers MakeModifiers(TypeKind typeKind, DiagnosticBag diag
...
+                        default:
+                            // No modifier token list
+                            modifierTokens = default;
+                            break;
+                    }
+
                     mods = ModifierUtils.CheckModifiers(
                         mods, allowedModifiers, declaration.Declarations[i].NameLocation, diagnostics,
-                        modifierTokensOpt: null, modifierErrors: out modifierErrors);
+                        modifierTokens: modifierTokens, modifierErrors: out modifierErrors);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Nit: could remove the `modifierTokens:` argument name.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs [hunk 2]
```diff
@@ -13,6 +13,24 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Parsing
...
+    partial abstract class A {}
+    partial abstract class A {}
+
+    partial partial class B {}
+    partial partial class B {}
+
+    partial abstract struct S {}
+    partial abstract struct S {}
+}");
+            comp.VerifyDiagnostics();
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: I didn't understand why there are no diagnostics. Shouldn't the CheckModifiers call report that partial is misplaced?

Comment (agocke) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: Ah, this should fail now. I may have forgotten to update the baseline.

**Pull request merged**